### PR TITLE
Support get contact points and get closest point on shape from a point

### PR DESCRIPTION
**Pull request by luzhuang**
### Please check if the PR fulfills these requirements

- [X] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## Summary by CodeRabbit

- **New Features**
	- Introduced the `ContactPoint` class for structured collision data.
	- Added `getClosestPoint` method to `ColliderShape` and its implementations for distance calculations.
	- Implemented `ICollision` interface to encapsulate collision event data.
	- Enhanced collision event handling in various physics classes to use the new `ICollision` parameter.
	- Added `pointDistance` method to `LiteBoxColliderShape` and `LiteSphereColliderShape` for distance calculations.
	- Introduced `pointDistance` method in `LiteDynamicCollider` for improved distance calculations.

- **Bug Fixes**
	- Improved error logging in `LiteDynamicCollider` by replacing `console.error` with `Logger.error`.

- **Tests**
	- Added new test suites for `ColliderShape` in both `LitePhysics` and `PhysX` contexts.
	- Established tests for collision functionality and validation of distance calculations.
	- Added unit tests for the new collision features and methods.

### CODE REVIEW

File: packages/core/src/physics/shape/ColliderShape.ts [hunk 1]
```diff
@@ -133,6 +135,27 @@ export abstract class ColliderShape implements ICustomClone {
...
   }
 
+  /**
+   * Get the distance and the closest point on the shape from a point.
+   * @param point - The point
+   * @param outClosestPoint - The result of the closest point on the shape
+   * @returns The distance between the point and the shape
+   */
+  getDistanceAndClosestPointFromPoint(point: Vector3, outClosestPoint: Vector3): number {
+    const tempQuat = ColliderShape._tempWorldRot;
```
Comment (zhuxudong) [inline packages/core/src/physics/shape/ColliderShape.ts:145]: should return input point and distance 0 if collider is not active

**Comment by zhuxudong** [comment]
support it also in lite?

File: packages/core/src/physics/shape/ColliderShape.ts [hunk 2]
```diff
@@ -133,6 +135,27 @@ export abstract class ColliderShape implements ICustomClone {
     Engine._physicalObjectsMap[this._id] = this;
   }
 
+  /**
+   * Get the distance and the closest point on the shape from a point.
+   * @param point - The point
+   * @param outClosestPoint - The result of the closest point on the shape
+   * @returns The distance between the point and the shape
+   */
+  getDistanceAndClosestPointFromPoint(point: Vector3, outClosestPoint: Vector3): number {
```
[the remark this answers is not in the document]
Comment (luzhuang) [reply packages/core/src/physics/shape/ColliderShape.ts:145]: done

File: packages/core/src/physics/shape/ColliderShape.ts (hunk 1 above)
Comment (luzhuang) [reply packages/core/src/physics/shape/ColliderShape.ts:151]: done

File: packages/core/src/physics/Collision.ts [hunk 3]
```diff
@@ -1,5 +1,58 @@
+import { Vector3 } from "@galacean/engine-math";
 import { ColliderShape } from "./shape";
+import { ICollision } from "@galacean/engine-design";
+
+/**
+ * Describes a contact point where the collision occurs.
+ */
+export interface ContactPoint {
```
Comment (GuoLei1990) [inline packages/core/src/physics/Collision.ts:8]: Use class

File: packages/core/src/physics/Collision.ts [hunk 4]
```diff
@@ -1,5 +1,58 @@
...
+  readonly normal: Vector3;
+  /** The impulse applied at the contact point, in world space. Divide by the simulation time step to get a force value. */
+  readonly impulse: Vector3;
+  /** The separation of the shapes at the contact point.  A negative separation denotes a penetration. */
+  readonly separation: number;
+}
 
 export class Collision {
+  /** @internal */
+  _nativeCollision: ICollision;
```
Comment (GuoLei1990) [inline packages/core/src/physics/Collision.ts:21]: Add comments for class `Collision`

File: packages/core/src/physics/Collision.ts [hunk 5]
```diff
@@ -1,5 +1,58 @@
...
+
+  /** The shape be collided. */
   shape: ColliderShape;
+
+  /**
+   * Get count of contact points.
+   */
```
Comment (GuoLei1990) [inline packages/core/src/physics/Collision.ts:28]: `Count of contact points.`

File: packages/physics-lite/src/shape/LiteBoxColliderShape.ts [hunk 6]
```diff
@@ -55,6 +57,48 @@ export class LiteBoxColliderShape extends LiteColliderShape implements IBoxColli
     this._setBondingBox();
   }
 
+  /**
+   * {@inheritDoc IColliderShape.pointDistance }
+   */
+  override pointDistance(position: Vector3, rotation: Quaternion, point: Vector3): IPointDistanceInfo {
```
Comment (zhuxudong) [inline packages/physics-lite/src/shape/LiteBoxColliderShape.ts:63]: why this api is not same as physics's `getDistanceAndClosestPointFromPoint`

File: packages/core/src/physics/Collision.ts [hunk 7]
```diff
@@ -1,5 +1,58 @@
...
+    return this._nativeCollision.contactCount;
+  }
+
+  /**
+   * Get contact points.
+   * @param outContacts - The result of contact points
+   * @returns The result of contact points
+   */
+  getContacts(outContacts: ContactPoint[]): ContactPoint[] {
+    const { shape0Id, shape1Id } = this._nativeCollision;
```
Comment (GuoLei1990) [inline packages/core/src/physics/Collision.ts:39]: Add `outContacts.length = 0` to clear

File: packages/physics-lite/src/shape/LiteBoxColliderShape.ts (hunk 6 above)
Comment (luzhuang) [reply packages/physics-lite/src/shape/LiteBoxColliderShape.ts:63]: pointDistance is  API name of physX, we can't pass pointer to native

File: packages/core/src/physics/Collision.ts [hunk 8]
```diff
@@ -1,5 +1,58 @@
...
+    const nativeContactPoints = this._nativeCollision.getContacts();
```
Comment (GuoLei1990) [inline packages/core/src/physics/Collision.ts:40]: If we can reduce GC?

File: packages/core/src/physics/shape/ColliderShape.ts [hunk 9]
```diff
@@ -133,6 +136,50 @@ export abstract class ColliderShape implements ICustomClone {
...
```
Comment (GuoLei1990) [inline packages/core/src/physics/shape/ColliderShape.ts:145]:
```ts
#include "PxPhysicsAPI.h"

using namespace physx;

PxVec3 getClosestPointOnShape(PxShape* shape, const PxVec3& point) {
    PxGeometryHolder geomHolder = shape->getGeometry(); // 获取 PxShape 的几何信息

    if (!geomHolder.any().isValid()) {
        throw std::runtime_error("Invalid geometry in shape.");
    }

    PxTransform pose = shape->getActor()->getGlobalPose() * shape->getLocalPose(); // 获取 PxShape 的全局变换

    PxVec3 closestPoint;
    PxReal distance = PxGeometryQuery::pointDistance(point, geomHolder.any(), pose, &closestPoint);

    return closestPoint; // 返回最近点
}

```

File: packages/design/src/physics/shape/IColliderShape.ts [hunk 10]
```diff
@@ -41,8 +41,31 @@ export interface IColliderShape {
...
   /**
    * Decrements the reference count of a shape and releases it if the new reference count is zero.
    */
   destroy(): void;
 }
+
+/**
+ * Distance information of a point to the shape.
+ */
+export interface IPointDistanceInfo {
```
Comment (GuoLei1990) [inline packages/design/src/physics/shape/IColliderShape.ts:61]: Can be merged into a Vector4. For internal interface specifications, appropriate trade-offs can be made.

File: packages/core/src/physics/Collision.ts [hunk 11]
```diff
@@ -1,5 +1,58 @@
...
+        factor = -1;
+      }
+      const contact: ContactPoint = {
+        position: new Vector3(position.x, position.y, position.z),
+        normal: new Vector3(normal.x, normal.y, normal.z).scale(factor),
+        impulse: new Vector3(impulse.x, impulse.y, impulse.z).scale(factor),
+        separation: separation
+      };
+      outContacts.push(contact);
+    }
```
Comment (GuoLei1990) [inline packages/core/src/physics/Collision.ts:55]: use `outContacts[i] ||= new ContactPoint()` mode

File: packages/core/src/physics/Collision.ts (hunk 4 above)
Comment (GuoLei1990) [inline packages/core/src/physics/Collision.ts:21]: What do you think if collision param is per Script?

Comment (luzhuang) [reply packages/core/src/physics/Collision.ts:24]: done

File: packages/core/src/physics/Collision.ts (hunk 5 above)
Comment (luzhuang) [reply packages/core/src/physics/Collision.ts:31]: done

File: packages/core/src/physics/Collision.ts (hunk 7 above)
Comment (luzhuang) [reply packages/core/src/physics/Collision.ts:44]: opted

File: packages/core/src/physics/Collision.ts (hunk 8 above)
Comment (luzhuang) [reply packages/core/src/physics/Collision.ts:45]: opted

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**